VX_RPP Extend OpenVX tensor support for augmentations in RPP

fiona-gladwin opened the pull request:
- Extend OpenVX tensor support for 26 augmentations.
- Minor code clean up in Brightness, CMN, Resize, Copy, Nop and SequenceRearrange.

fiona-gladwin committed:
Restructure changes

fiona-gladwin committed:
Add tensor support for remaining augmentations

fiona-gladwin committed:
Code clean up and fixes

Brightness, CMN, Resize, Copy and Nop

fiona-gladwin committed:
Code clean up and fixes

Exposure ColorTemp Contrast Blur GammaCorrection

fiona-gladwin committed:
Code clean up and fixes

Vignette Jitter WarpAffine Rotate Flip

fiona-gladwin committed:
Code clean up and fixes

Pixelate Rain ResizeCrop ResizeCropNormalize ResizeMirrorNormalize

fiona-gladwin committed:
Code clean up and fixes

Snow Hue Crop ColorTwist Saturation

fiona-gladwin committed:
Resolve build errors

fiona-gladwin committed:
Code clean up and fixes

Blend FishEye Fog Glitch LensCorrection

fiona-gladwin committed:
Code clean up and fixes

Noise and SequenceRearrange

fiona-gladwin committed:
Minor code changes

fiona-gladwin committed:
Minor fix warnings

fiona-gladwin committed:
Minor fix

fiona-gladwin committed:
Change function name for tensor augmentation nodes

fiona-gladwin committed:
Fix build issues wrt Sequence rearrange

Reintroduce the vx_image support
Rename to SequenceRearrangeImage for file with image support

fiona-gladwin committed:
Code clean up - SequenceRearrange

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_vx_tensor_augmentations

fiona-gladwin committed:
Merge branch 'lk/rocalTensor_build_fix' into PR_vx_tensor_augmentations

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Fix ResizeCrop

fiona-gladwin committed:
Minor changes - Remove void * casting

Add space after delete operator and if condition

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Code clean up

Blend, Blur, ColorTemp and ColorTwist

fiona-gladwin committed:
Pass input and output layout to ColorTwist

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Code clean up vx_rpp

Minor changes in variable names
Change data type to Rpp type

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into lk/rocalTensor_build_fix

fiona-gladwin committed:
Code clean up - VX tensor augmentations

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Minor changes

Correct spellings, variable name

fiona-gladwin committed:
Rearrange augmentation names in Alphabetical order

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Add constant def for WarpAffine

fiona-gladwin committed:
Change description to include data type

fiona-gladwin committed:
Rename sequence rearrange image to batchPD

fiona-gladwin committed:
Add description for all tensor functions in vx_rpp

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Merge branch 'lk/rocalTensor_build_fix' into PR_vx_tensor_augmentations

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into lk/rocalTensor_build_fix

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_vx_tensor_augmentations

fiona-gladwin committed:
Minor change - Vignette

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Minor change

fiona-gladwin committed:
Merge branch 'lk/rocalTensor_build_fix' into PR_vx_tensor_augmentations

fiona-gladwin committed:
Add prefix vxExtRpp to VX_RPP tensor augmentation functions

fiona-gladwin committed:
Merge branch 'lk/rocalTensor_build_fix' into PR_vx_tensor_augmentations

fiona-gladwin committed:
Change prefix to vxExtRpp for tensor augmentations vx_rpp

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/include/internal_publishKernels.h:
```diff
@@ -209,14 +236,41 @@ vx_status Resize_Register(vx_context);
 #define VX_KERNEL_RPP_CROPPD_NAME   							"org.rpp.CropPD"
```
@fiona-gladwin Please check indentation here and below

fiona-gladwin replied:
@LakshmiKumar23 I am not able to see any indentation issues when I check in VScode please let me know which line the issue is

LakshmiKumar23 replied:
@fiona-gladwin Please open the file on git and check. It might be space vs tab and looks same on vs code

fiona-gladwin replied:
In git file also I am unable to see the issue

LakshmiKumar23 replied:
@fiona-gladwin. Refer below. It's all over the place
![image](https://github.com/LakshmiKumar23/MIVisionX/assets/27569995/4fbce97b-59ee-40b3-bc39-2f4a258a281b)

fiona-gladwin replied:
Done. Please check

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -982,9 +982,11 @@ extern "C"
 	 * \param [in] graph The handle to the graph.
-	 * \param [in] pSrc The input tensor data.
+	 * \param [in] pSrc The input tensor in <tt>\ref VX_TYPE_UINT8<tt> or <tt>\ref VX_TYPE_FLOAT32<tt> or 
+	 * <tt>\ref VX_TYPE_FLOAT16<tt> or <tt>\ref VX_TYPE_INT8<tt> format data.
```
@fiona-gladwin why the extra line *. This might cause issues with Doxygen. Keep all comments for on parameter as one continuous line

fiona-gladwin replied:
Done. changed into single line

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -982,9 +982,11 @@ extern "C"
 	 * \param [in] pSrcRoi The input tensor of batch size in <tt>unsigned int<tt> containing the roi values for the input in xywh/ltrb format.
-	 * \param [out] pDst The output tensor data.
+	 * \param [out] pDst The output tensor in <tt>\ref VX_TYPE_UINT8<tt> or <tt>\ref VX_TYPE_FLOAT32<tt> or 
+	 * <tt>\ref VX_TYPE_FLOAT16<tt> or <tt>\ref VX_TYPE_INT8<tt> format data.
```
@fiona-gladwin same as above. I see many more like this below too

fiona-gladwin replied:
Done

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -1005,27 +1007,29 @@ extern "C"
+	SHARED_PUBLIC vx_node VX_API_CALL vxExtRppCropMirrorNormalize(vx_graph graph, vx_tensor pSrc, vx_tensor pSrcRoi, vx_tensor pDst, vx_array pMultiplier, vx_array pOffset, vx_array pMirror, vx_scalar inputLayout, vx_scalar outputLayout, vx_scalar roiType);
 
-	/*! \brief [Graph] Creates a RPP Nop function node.
+	/*! \brief [Graph] Creates a tensor Nop function node.
```
@fiona-gladwin Some of these have been changed from "RPP" to "tensor. Others haven't. Please be consistent

fiona-gladwin replied:
The Nop node does not call any RPP functions internally, same applies for copy and sequence rearrange hence we have named it as tensor Nop function node

rrawther replied:
Yes. Still we want the name and arguments to be consistant all throughout the API. The user of vx_rpp doesn't really care whether it calls a low level RPP or not

fiona-gladwin replied:
Done

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    data->pSrcRoi = reinterpret_cast<RpptROI *>(roi_tensor_ptr);
+    // Fill width and height array with ROI data required by RPP batchPD kernels
```
@fiona-gladwin WHy is this going to call batch PD kernel?

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp:
```diff
@@ -0,0 +1,259 @@
+        return_status = VX_ERROR_NOT_IMPLEMENTED;
+#elif ENABLE_HIP
+        if (data->pSrcDesc->c == 1) {
+            rpp_status = rppi_blur_u8_pln1_batchPD_gpu(data->pSrc, data->pSrcDimensions, data->maxSrcDimensions, data->pDst, data->pKernelSize, data->pSrcDesc->n, data->handle->rppHandle);
```
@fiona-gladwin why is it calling rppi function? isn't rppi = image? rppt = tensor?

fiona-gladwin replied:
Yes, but for certain augmentations like Blur, snow, rain there is no rppt calls present in RPP. So internally we will be calling the batchPD kernels for these augmentations

For augmentations like Pixelate, Vignette, Jitter and ColorTemperature the support is added for rppt but the codes are not pushed to TOT so it is commented for now

LakshmiKumar23 commented on amd_openvx_extensions/amd_rpp/source/tensor/FishEye.cpp:
```diff
@@ -0,0 +1,252 @@
+/*
+Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
@fiona-gladwin All copyright needs to be updated to say 2023, not 2019-2023

fiona-gladwin replied:
Done

fiona-gladwin committed:
change copy right year in tensor augmentations

fiona-gladwin committed:
change unsigned to int in for loop

rrawther commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -994,9 +996,9 @@ extern "C"
 	 * \returns A node reference <tt>\ref vx_node</tt>. Any possible errors preventing a
```
Combine line #995 and #996 into one. They are both saying the same thing and will be better for doxygen.  This is true for all the fuctions.

LakshmiKumar23 replied:
Yes, as mentioned in [this](https://github.com/LakshmiKumar23/MIVisionX/pull/32/files/897577f1145788898b0343d204be6e99728df33a#r1280922686) too

fiona-gladwin replied:
Done

rrawther commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -1005,27 +1007,29 @@ extern "C"
+	SHARED_PUBLIC vx_node VX_API_CALL vxExtRppCropMirrorNormalize(vx_graph graph, vx_tensor pSrc, vx_tensor pSrcRoi, vx_tensor pDst, vx_array pMultiplier, vx_array pOffset, vx_array pMirror, vx_scalar inputLayout, vx_scalar outputLayout, vx_scalar roiType);
 
-	/*! \brief [Graph] Creates a RPP Nop function node.
+	/*! \brief [Graph] Creates a tensor Nop function node.
```
please remove tensor prefix here

fiona-gladwin replied:
Done

rrawther commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -1034,25 +1038,528 @@ extern "C"
+	SHARED_PUBLIC vx_node VX_API_CALL vxExtRppNop(vx_graph graph, vx_tensor pSrc, vx_tensor pDst);
 
-	/*! \brief [Graph] Creates a RPP vxRppResize function node.
+	/*! \brief [Graph] Creates a RPP Resize function node.
```
please remove RPP prefix here

fiona-gladwin replied:
Done

rrawther commented on amd_openvx_extensions/amd_rpp/include/vx_ext_rpp.h:
```diff
@@ -1034,25 +1038,528 @@ extern "C"
+	 */
+	SHARED_PUBLIC vx_node VX_API_CALL vxExtRppSequenceRearrange(vx_graph graph, vx_tensor pSrc, vx_tensor pDst, vx_array pNewOrder, vx_scalar layout);
+
+	/*! \brief [Graph] Creates a RPP Blend function node.
```
I think we need to remove RPP prefix from all the descriptions since it is implied. We don't want to describe whether it is calling low level function or not

fiona-gladwin replied:
Removed for all tensor API

rrawther commented on amd_openvx_extensions/amd_rpp/source/image/SequenceRearrangebatchPD.cpp:
```diff
@@ -0,0 +1,294 @@
+    vx_uint32 sequence_count;
+    vx_uint32 *new_order;
+#if ENABLE_OPENCL
+    cl_mem cl_pSrc;
```
I think I have commented on this before. We shouldn't mix camel and snake_case both. Please keep everything in single type

fiona-gladwin replied:
This is a batchPD file. In our previous PR we made tensor changes in the batchPD code itself. So we have have reintroduced the file and changed name alone. In all our tensor files we have followed the naming convention in all files

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_vx_tensor_augmentations

fiona-gladwin committed:
Modified all tensor API descriptions

fiona-gladwin commented:
@rrawther and @LakshmiKumar23 Addressed all review comments
As discussed. Modified the Doxygen description of all tensor API in this PR.

LakshmiKumar23 commented:
> @rrawther and @LakshmiKumar23 Addressed all review comments As discussed. Modified the Doxygen description of all tensor API in this PR.

@fiona-gladwin did the document build successfully?

fiona-gladwin commented:
> > @rrawther and @LakshmiKumar23 Addressed all review comments As discussed. Modified the Doxygen description of all tensor API in this PR.
> 
> @fiona-gladwin did the document build successfully?

Yes @LakshmiKumar23 I was able to build the documentation, but with multiple warnings
```
====================== slowest reading durations =======================
0.610 docs/docs/docs/README
0.539 README
0.534 docs/README
0.532 docs/docs/README
0.287 docs/docs/CHANGELOG
build succeeded, 2821 warnings.

The HTML pages are in _build/html.
```

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp:
```diff
@@ -0,0 +1,243 @@
+    data->pDstDesc->offsetInBytes = 0;
+    fillDescriptionPtrfromDims(data->pDstDesc, data->outputLayout, data->ouputTensorDims);
+
+    data->pAlpha = static_cast<Rpp32f *>(malloc(sizeof(Rpp32f) * data->pSrcDesc->n));
```
please use new  and delete instead

fiona-gladwin replied:
Done, changed in all files

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/Blend.cpp:
```diff
@@ -0,0 +1,243 @@
+static vx_status VX_CALLBACK uninitializeBlend(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    BlendLocalData *data;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    if (data->pAlpha != nullptr)  free(data->pAlpha);
```
see comment above

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/Blur.cpp:
```diff
@@ -0,0 +1,259 @@
+
+    data->maxSrcDimensions.height = data->pSrcDesc->h;
+    data->maxSrcDimensions.width = data->pSrcDesc->w;
+    data->pKernelSize = static_cast<Rpp32u *>(malloc(sizeof(Rpp32u) * data->pSrcDesc->n));
```
please use new instead of malloc whereever applicable

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/Brightness.cpp:
```diff
@@ -27,8 +27,8 @@ struct BrightnessLocalData {
     RppPtr_t pDst;
-    vx_float32 *pAlpha;
-    vx_float32 *pBeta;
+    Rpp32f *pAlpha;
```
why is this changed? It is better to use vx_ types for vx_extensions

fiona-gladwin replied:
Most of the members in this structure makes use of Rpp struct definitions and types. Hence in order to maintain uniformity and for better code we have changed all members to be of Rpp type. Please let us know if we can retain the same

fiona-gladwin replied:
Done

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/ColorTemperature.cpp:
```diff
@@ -0,0 +1,240 @@
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[5], &output_layout, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[6], &roi_type, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[7], &data->deviceType, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    data->roiType = (roi_type == 0) ? RpptRoiType::XYWH : RpptRoiType::LTRB;
```
you can use better code here. Can't you pass the same value as enum for roiType and do a cast here?

fiona-gladwin replied:
Done, changed in all files

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp:
```diff
@@ -0,0 +1,257 @@
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[8], &output_layout, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[9], &roi_type, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    STATUS_ERROR_CHECK(vxCopyScalar((vx_scalar)parameters[10], &data->deviceType, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
+    data->roiType = (roi_type == 0) ? RpptRoiType::XYWH : RpptRoiType::LTRB;
```
please se the comment in  previous file. Applies to all

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp:
```diff
@@ -0,0 +1,257 @@
+
+static vx_status VX_CALLBACK refreshColorTwist(vx_node node, const vx_reference *parameters, vx_uint32 num, ColorTwistLocalData *data) {
+    vx_status status = VX_SUCCESS;
+    STATUS_ERROR_CHECK(vxCopyArrayRange((vx_array)parameters[3], 0, data->inputTensorDims[0], sizeof(Rpp32f), data->pAlpha, VX_READ_ONLY, VX_MEMORY_TYPE_HOST));
```
I have mentioned this before. We need to pass this as a single tensor ptr with 4 float values for (alpha, beta, hue and sat). This was we can pass the same ptr (host/device) to RPP without copying

fiona-gladwin replied:
Hi @rrawther 
We discussed about the changes required for this in call and we decided not to do this.
Below were the pointers discussed.
Currently we have all of these variables as vx_array in rocAL as they are generated using the Parameter class. Internally it accepts batch size and generates random or fixed values and fills the vx_array.
In order to convert to vx_tensor we need to create a tensor fill with all the array values and again at the end when we need to pass to RPP we must split it into the 4 arrays.
Also in cases where we have multiple inputs with different data types we might not be able to do it.

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/ColorTwist.cpp:
```diff
@@ -0,0 +1,257 @@
+    data->pDstDesc->offsetInBytes = 0;
+    fillDescriptionPtrfromDims(data->pDstDesc, data->outputLayout, data->outputTensorDims);
+
+    data->pAlpha = static_cast<Rpp32f *>(malloc(sizeof(Rpp32f) * data->pSrcDesc->n));
```
we need to pass the values directly from rocAL to rpp without copying as a single tensor

rrawther replied:
Please make a not of this to do it as a separate PR

fiona-gladwin committed:
Replace malloc and free with new and delete operator

fiona-gladwin committed:
Change ROI type condition to static cast

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/CropMirrorNormalize.cpp:
```diff
@@ -27,9 +27,9 @@ struct CropMirrorNormalizeLocalData {
-    vx_float32 *pMultiplier;
-    vx_float32 *pOffset;
-    vx_uint32 *pMirror;
+    Rpp32f *pMultiplier;
```
I think it is better to use vx_types for OpenVX. Also we need to pass parameters through a single tensor instead of multiple vx_arrays

fiona-gladwin replied:
Changed to vx_types

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/Fog.cpp:
```diff
@@ -0,0 +1,259 @@
+static vx_status VX_CALLBACK uninitializeFog(vx_node node, const vx_reference *parameters, vx_uint32 num) {
+    FogLocalData *data;
+    STATUS_ERROR_CHECK(vxQueryNode(node, VX_NODE_LOCAL_DATA_PTR, &data, sizeof(data)));
+    if (data->pFogValue != nullptr)  free(data->pFogValue);
```
please use new and delete whereever possible

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/Glitch.cpp:
```diff
@@ -0,0 +1,270 @@
+    Rpp32u deviceType;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    Rpp32u *pXOffsetR;
```
We need to combine all these offsets into a single tensor parameter

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/LensCorrection.cpp:
```diff
@@ -0,0 +1,265 @@
+    Rpp32u deviceType;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    Rpp32f *pStrength;
```
I see this in all the kernels where parameters are passed as arrays which needs extra copies and reformatting in vx_rpp

rrawther commented on amd_openvx_extensions/amd_rpp/source/tensor/Rain.cpp:
```diff
@@ -0,0 +1,277 @@
+    Rpp32u deviceType;
+    RppPtr_t pSrc;
+    RppPtr_t pDst;
+    Rpp32f *pRainValue;
```
same as comment before

rrawther reviewed (changes requested):
I think we need to change how the parameters are passed from rocAL to vx_rpp. Combining all the parameters into one tensor and using pinned memory will make it more efficient. I have discussed this before. But all the code is still using the batch_pd style

fiona-gladwin committed:
Merge branch 'lk/rocalTensor' of https://github.com/LakshmiKumar23/MIVisionX into PR_vx_tensor_augmentations

fiona-gladwin committed:
Change data type from Rpp type to vx type

fiona-gladwin commented:
Changed all RPP types to vx data types.
As discussed will work on the vx_tensor changes in the future

LakshmiKumar23 committed:
Merge branch 'lk/rocalTensor' into PR_vx_tensor_augmentations

fiona-gladwin committed:
Fix indentation in internalPublishkernels.h

fiona-gladwin committed:
Merge branch 'PR_vx_tensor_augmentations' of https://github.com/fiona-gladwin/MIVisionX into PR_vx_tensor_augmentations

LakshmiKumar23 commented on rocAL/rocAL/include/api/rocal_api_types.h:
```diff
@@ -177,8 +177,8 @@ enum RocalTensorBackend {
-    ROCAL_XYWH = 0,
-    ROCAL_LTRB = 1
+    ROCAL_LTRB = 0,
+    ROCAL_XYWH = 1
```
@fiona-gladwin Why is this interchanged? Has this been reflected in all use cases?

fiona-gladwin replied:
https://github.com/LakshmiKumar23/MIVisionX/pull/32#discussion_r1286480719
With respect to the above comment, We need to do a typecast from the RocalROIType to the RpptROIType directly
Hence the order should be the same as that of RPP. Hence this change is made

LakshmiKumar23 commented on rocAL/rocAL/include/pipeline/commons.h:
```diff
@@ -106,8 +106,8 @@ enum class DecodeMode
-    LTRB
+    LTRB = 0,
+    XYWH
 };
```
@fiona-gladwin same as abaove?

LakshmiKumar23 commented:
@rrawther @fiona-gladwin ready to merge?
